### PR TITLE
[TEP-0076]Support Results Array Indexing

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
# Changes

This is part of work in TEP-0076.
This commit provides the support to refer array indexing results.
Previous this commit we support emitting array results so users can
write array results to task level, but we cannot pass array results via
index between tasks within one pipeline. This commit adds the support for this.

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes
```release-note
Support indexing array results substitution as an alpha feature.

A task can specify a type to produce array result, such as:

  results:
    - name: array-results
       type: array
       description: The array results

And the task script can populate result in an array form with:

echo -n "[\"hello\",\"world\"]" | tee $(results.array-results.path)

and we can refer to the array results elements via index in param like:
  params:
    - name: foo
      value: "$(tasks.task1.results.array-results[1])"

This feature is part of the TEP-0076. 
```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/apply_test.go [hunk 1]
```diff
@@ -401,6 +401,36 @@ func TestApplyTaskResults_MinimalExpression(t *testing.T) {
 				}},
 			},
 		}},
+	}, {
```
Comment (ywluogg) [inline pkg/reconciler/pipelinerun/resources/apply_test.go:404]: Can you add several more test cases to test invalid and out of index cases?

File: pkg/apis/pipeline/v1beta1/resultref.go [hunk 2]
```diff
@@ -35,13 +35,17 @@ const (
 	// ResultResultPart Constant used to define the "results" part of a pipeline result reference
 	ResultResultPart = "results"
 	// TODO(#2462) use one regex across all substitutions
-	variableSubstitutionFormat = `\$\([_a-zA-Z0-9.-]+(\.[_a-zA-Z0-9.-]+)*\)`
+	// variableSubstitutionFormat matches format like $result.resultname, $result.resultname[int] and $result.resultname[*]
+	variableSubstitutionFormat = `\$\([_a-zA-Z0-9.-]+(\.[_a-zA-Z0-9.-]+)*(\[([0-9])*\*?\])?\)`
+	// excludeArrayIndexing will replace all `[int]` and `[*]` for parseExpression to extract result name
+	excludeArrayIndexing = `\[([0-9])*\*?\]`
```
Comment (ywluogg) [inline pkg/apis/pipeline/v1beta1/resultref.go:41]: Discussed offline about renaming

File: pkg/reconciler/pipelinerun/resources/apply_test.go [hunk 3]
```diff
@@ -479,6 +541,36 @@ func TestApplyTaskResults_EmbeddedExpression(t *testing.T) {
...
+		name: "Test array indexing result substitution on embedded variable substitution expression - params",
```
Comment (ywluogg) [inline pkg/reconciler/pipelinerun/resources/apply_test.go:545]: Same as above for adding invalid test cases

**Review by ywluogg — COMMENTED** [review]
Mostly LGTM. Please fix those comments for test cases

File: docs/variables.md [hunk 4]
```diff
@@ -22,6 +22,9 @@ For instructions on using variable substitutions see the relevant section of [th
 | `tasks.<taskName>.results.<resultName>` | The value of the `Task's` result. Can alter `Task` execution order within a `Pipeline`.) |
 | `tasks.<taskName>.results['<resultName>']` | (see above)) |
 | `tasks.<taskName>.results["<resultName>"]` | (see above)) |
+| `tasks.<taskName>.results.<resultName>[i]` | The ith value of the `Task's` array result. Can alter `Task` execution order within a `Pipeline`.) |
+| `tasks.<taskName>.results['<resultName>'][i]` | (see above)) |
+| `tasks.<taskName>.results["<resultName>"][i]` | (see above)) |
```
Comment (Yongxuanzhang) [inline docs/variables.md:27]: These two are not supported because of previous regex bug. We may remove this and add them back when it is supported

**Review by vdemeester — APPROVED** [review]

File: pkg/reconciler/pipelinerun/resources/apply_test.go (hunk 1 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/pipelinerun/resources/apply_test.go:404]: I added the validation in the latest commit,

Comment (ywluogg) [reply pkg/reconciler/pipelinerun/resources/apply_test.go:404]: LGTM. You can resolve this

**Comment by Yongxuanzhang** [comment]
> Can you also add examples into the release notes?

Yes! I will add release notes soon

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+metadata:
+  name: pipelinerun-array-indexing-results
+spec:
+  pipelineSpec:
+    tasks:
+      - name: task1
+        taskSpec:
+          results:
+            - name: array-results
+              description: The array resluts
```
Comment (lbernick) [inline examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml:12]: there's a typo here and in the name of this file

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 6]
```diff
@@ -306,8 +306,10 @@ func (t *ResolvedPipelineRunTask) skipBecauseResultReferencesAreMissing(facts *P
 	if t.checkParentsDone(facts) && t.hasResultReferences() {
 		resolvedResultRefs, pt, err := ResolveResultRefs(facts.State, PipelineRunState{t})
 		rprt := facts.State.ToMap()[pt]
-		if err != nil && (t.IsFinalTask(facts) || rprt.Skip(facts).SkippingReason == v1beta1.WhenExpressionsSkip) {
-			return true
+		if rprt != nil {
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:309]: what's the reason for this change? it doesn't look related to results

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml (hunk 5 above)
Comment (pritidesai) [reply examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml:12]: isn't specifying `type` necessary here?

File: docs/variables.md [hunk 7]
```diff
@@ -22,6 +22,9 @@ For instructions on using variable substitutions see the relevant section of [th
 | `tasks.<taskName>.results.<resultName>` | The value of the `Task's` result. Can alter `Task` execution order within a `Pipeline`.) |
 | `tasks.<taskName>.results['<resultName>']` | (see above)) |
 | `tasks.<taskName>.results["<resultName>"]` | (see above)) |
+| `tasks.<taskName>.results.<resultName>[i]` | The ith value of the `Task's` array result. Can alter `Task` execution order within a `Pipeline`.) |
```
Comment (pritidesai) [inline docs/variables.md:25]: How about start notation - `tasks.<taskName>.results.<resultName>[*]`? is that something supported in this context?

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml (hunk 5 above)
Comment (Yongxuanzhang) [reply examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml:12]: > there's a typo here and in the name of this file

Thank you!

> isn't specifying type necessary here?
Yes, I will add this, thanks for catching! It is not failing in this pr because the validation we discussed is implemented in this pr https://github.com/tektoncd/pipeline/pull/4920

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+          steps:
+            - name: print-param
+              image: ubuntu
+              script: |
+                #!/bin/bash
+                VALUE=$(params.foo)
+                EXPECTED="2"
+                diff=$(diff <(printf "%s\n" "${VALUE[@]}") <(printf "%s\n" "${EXPECTED[@]}"))
+                if [[ -z "$diff" ]]; then
+                    echo "TRUE"
```
[the remark this answers is not in the document]
Comment (Yongxuanzhang) [reply examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml:37]: Oh thanks! Learn something new today. 😄

File: examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml (hunk 5 above)
Comment (lbernick) [reply examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml]: filename still needs to be fixed

Comment (Yongxuanzhang) [reply examples/v1beta1/pipelineruns/alpha/pipelinerun-array-resluts-indexing.yaml]: Oh sorry my bad!

**Comment by Yongxuanzhang** [comment]
> happy to LGTM but @pritidesai just want to make sure your concerns are addressed first

Yeah sure! I will wait for @pritidesai's comments

**Comment by ywluogg** [comment]
Link this to #4723

File: pkg/apis/pipeline/v1beta1/resultref.go [hunk 9]
```diff
@@ -122,12 +129,19 @@ func stripVarSubExpression(expression string) string {
...
 	}
-	return subExpressions[1], subExpressions[3], nil
+
+	stringIdx := strings.TrimSuffix(strings.TrimPrefix(arrayIndexingRegex.FindString(subExpressions[3]), "["), "]")
+	subExpressions[3] = arrayIndexingRegex.ReplaceAllString(subExpressions[3], "")
+	if stringIdx != "" {
+		intIdx, _ := strconv.Atoi(stringIdx)
+		return subExpressions[1], subExpressions[3], intIdx, nil
+	}
+	return subExpressions[1], subExpressions[3], 0, nil
```
Comment (chuangw6) [inline pkg/apis/pipeline/v1beta1/resultref.go:144]: When the index part in bracket for array or the whole expression is invalid, or it's a reference to a string result, wouldn't it make more sense if we set the third return value to be `-1` to indicate it's not array or invalid value for index?
